### ISSUE TITLE
Output widget loses some output

### ISSUE THREAD

**Issue by krey** [jupyter-widgets/ipywidgets#2339]
I'm using version 7.4.2

```python
import ipywidgets as widgets

output = widgets.Output()
field = widgets.Text()
updating = False

def update(change):
    global updating
    with output:
        if not updating:
            updating = True
            print('updating value')
            field.value = 'a' + change.new
            print('finished updating value')
            updating = False
            
field.observe(update, names='value')
field.value = 'b'
```

After running this, the `Output` widget only contains `updating value`. The text `finished updating value` is printed in the output area below the notebook cell.

To fix it, I can add an additional context

```python
def update(change):
    global updating
    with output:
        if not updating:
            updating = True
            print('updating value')
            field.value = 'a' + change.new
            with output:
                print('finished updating value')
            updating = False
```

Here, the `Output` widget contains
```
updating value
finished updating value
```
as expected.

**Comment by maartenbreddels** [comment]
> ```python
> field.value = 'a' + change.new
> ```

This triggers update again right? which leads to a kind of nested `with output`, when the second update returns (after that line), the first context manager/ouput is not used anymore. I guess it's a big with nested outputs as context managers, if supported at all.

**Comment by krey** [comment]
@maartenbreddels thanks for taking a look.

This is a fairly common pattern for me:

Sanitizing user input
user types something into view -> sanitize and put into model -> update the view

Should I be doing something differently?

**Comment by krey** [comment]
Hmm, this seems to work

```python
@contextmanager
def output_widget_context(widget):
    if widget.msg_id == '':
        with widget:
            yield
    else:
        yield
```

**Comment by jasongrout** [comment]
> I guess it's a big with nested outputs as context managers, if supported at all.

Does #2384 solve this?

**Comment by SylvainCorlay** [comment]
This appears to be solved in #2384. Closing.
